independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/BeanTypeAssignabilityRules.java

```diff
@@ -43,13 +43,24 @@ private boolean matchesNoBoxing(Type requiredType, Type beanType) {
             if (beanType instanceof Class<?>) {
                 return matches((Class<?>) requiredType, (Class<?>) beanType);
             }
+            //Needs to precede the following check for ParameterizedType and technically redundant
+            //(as ParameterizedTypeImpl also implements ParameterizedType)
+            //but we place it here to avoid type cache pollution (see JDK-8180450)
+            if (beanType instanceof ParameterizedTypeImpl) {
+                return matches((Class<?>) requiredType, (ParameterizedTypeImpl) beanType);//N.B. to cast to the implementation
+            }
             if (beanType instanceof ParameterizedType) {
                 return matches((Class<?>) requiredType, (ParameterizedType) beanType);
             }
         } else if (requiredType instanceof ParameterizedType) {
```

Comment:
Seems funny that this one doesn't hurt :-)

Reply:
That's odd indeed. In my particular test this one doesn't show, but I'll improve the patch a bit so that it works better in other scenarios. Thanks for spotting :)